### PR TITLE
Added Charger object support to SDK

### PR DESCRIPTION
**Pull request by MarkWesleyAtAnki**
Added charger factory and object
Added docsstrings for factory objects in world.py
Send RequestAvailableObjects on robot initialize - and handle AvailableObjects message that is returned, this gives us info on connected cubes at the start instead of having to wait to visually see them.
Upgraded "Tap event received for unknown object ID" back to a warning as we should always know about that cube now
Fixed robot.py factory docstrings (they were missing a colon before class)
Made robot keep track of it's origin_id for poses
Added GoToObject support for driving to within X mm of a cube or charger
Added go_to_object_test.py example that shows usage of GoToObject action for finding a cube and driving up to it
Added drive_to_charger_test.py example that shows usage of accessing the charger and driving near to it.

### CODE REVIEW

File: examples/drive_to_charger_test.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import asyncio
+import sys
+import time
+
+import cozmo
+from cozmo.util import degrees, distance_mm, speed_mmps
+
+
+def find_charger(robot):
+    '''The core of the follow_faces program'''
```
Comment (msintov) [inline examples/drive_to_charger_test.py]: Do we want to keep this 'follow_faces' docstring?

File: src/cozmo/objects.py [hunk 2]
```diff
@@ -95,6 +95,17 @@ class EvtObjectAppeared(event.Event):
     pose = 'The cozmo.util.Pose defining the position and rotation of the object.'
 
 
+class EvtObjectAvailable(event.Event):
+    '''Triggered when the engine reports that an object is available (i.e. exists)
```
Comment (msintov) [inline src/cozmo/objects.py]: Need a period at the end of this first line.

File: src/cozmo/objects.py [hunk 3]
```diff
@@ -95,6 +95,17 @@ class EvtObjectAppeared(event.Event):
...
+
+    This will usually occur at the start of the program in response to the SDK
+    sending RequestAvailableObjects to the engine.
+    '''
+    obj = 'The object that is available'
+    updated = 'A set of field names that have changed'
+    pose = 'The cozmo.util.Pose defining the position and rotation of the object.'
```
Comment (msintov) [inline src/cozmo/objects.py]: The strings assigned to obj and updated don't end with periods but pose does. Should the period at the end of the pose string be removed?

File: src/cozmo/objects.py [hunk 4]
```diff
@@ -365,6 +398,12 @@ def set_lights_off(self):
         self.set_lights(lights.off_light)
 
 
+class Charger(ObservableObject):
+    '''Cozmo's charger object, can be seen and driven towards'''
```
Comment (msintov) [inline src/cozmo/objects.py]: Maybe reword: "Cozmo's charger object, which the robot can see and drive toward."

File: src/cozmo/robot.py [hunk 5]
```diff
@@ -1141,6 +1172,28 @@ def go_to_pose(self, pose, relative_to_robot=False):
         self._action_dispatcher._send_single_action(action)
         return action
 
+    def go_to_object(self, object, distance_from_object):
+        '''Tells Cozmo to drive to the specified object.
+
+        Args:
+            object (:class:`cozmo.objects.ObservableObject`): The destination object.
+            distance_from_object (:class:`cozmo.util.Distance`): The distance from
```
Comment (msintov) [inline src/cozmo/robot.py]: Maybe reword the distance_from_object description like this:
'The distance from the object to stop. This is the distance between the origins. For instance, the distance from the robot's origin (between Cozmo's two front wheels) to the cube's origin (at the center of the cube) is ~40mm.

File: src/cozmo/world.py [hunk 6]
```diff
@@ -136,7 +154,17 @@ def _allocate_object_from_msg(self, msg):
...
             return cube
 
-        elif (msg.objectFamily == _clad_to_game_cozmo.ObjectFamily.CustomObject):
+        elif msg.objectFamily == _clad_to_game_cozmo.ObjectFamily.Charger:
+            charger = self.charger_factory(self.conn, self, msg.objectID, dispatch_parent=self)
+            if self.charger:
+                logger.error('Allocating multiple chargers: existing charger=%s msg=%s', self.charger, msg)
+            self.charger = charger
+            self._objects[charger.object_id] = charger
+            charger._robot = self.robot  # XXX this will move if/when we have multi-robot support
```
Comment (msintov) [inline src/cozmo/world.py]: Should 'XXX' be a TODO instead?

File: examples/drive_to_charger_test.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            charger = robot.world.wait_for_observed_charger(timeout=30)
+            print("Found charger: %s" % charger)
+        except asyncio.TimeoutError:
+            print("Didn't see the charger")
+        finally:
+            # whether we find it or not, we want to stop the behavior
+            look_around.stop()
+
+    if not charger:
+        charger = robot.world.charger
```
Comment (gwatts) [inline examples/drive_to_charger_test.py]: why would `charger` be `None` at this point, yet `robot.world.charger` be set?

File: examples/drive_to_charger_test.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            else:
+                print("Attempting to drive to last known location of the charger")
+
+    if charger:
+        # Attempt to drive near to the charger, and then stop.
+        action = robot.go_to_object(charger, distance_mm(65.0))
+        action.wait_for_completed()
+        print("Completed action: result = %s" % action)
+        print("Done.")
+
```
Comment (gwatts) [inline examples/drive_to_charger_test.py]: all this `if charger` stuff feels a bit convoluted for an example

File: src/cozmo/objects.py [hunk 9]
```diff
@@ -186,6 +197,28 @@ def _observed_timeout(self):
         self._is_visible = False
         self.dispatch_event(EvtObjectDisappeared, obj=self)
 
+    def _handle_available_object(self, available_object):
+        # triggered when engine sends available objects
+        # as a response to a RequestAvailableObjects message
+        if self.last_observed_robot_timestamp and \
+                (self.last_observed_robot_timestamp > available_object.lastObservedTimestamp):
```
Comment (gwatts) [inline src/cozmo/objects.py]: Better style to wrap the whole if condition in parentheses and drop the backslash at the end of the first line

File: src/cozmo/objects.py [hunk 10]
```diff
@@ -186,6 +197,28 @@ def _observed_timeout(self):
...
+
+        self._pose = util.Pose(available_object.pose.x, available_object.pose.y, available_object.pose.z,
+                               q0=available_object.pose.q0, q1=available_object.pose.q1,
+                               q2=available_object.pose.q2, q3=available_object.pose.q3,
+                               origin_id=available_object.pose.originID)
+
+        self.dispatch_event(EvtObjectAvailable,
+                            obj=self,
+                            updated=changed_fields,
+                            pose = self._pose)
```
Comment (gwatts) [inline src/cozmo/objects.py]: stray whitespace around the `=`

File: src/cozmo/objects.py (hunk 4 above)
Comment (gwatts) [inline src/cozmo/objects.py]: should we stick with "observe" rather than "see" for the sake of consistency?

File: src/cozmo/robot.py [hunk 11]
```diff
@@ -1141,6 +1172,28 @@ def go_to_pose(self, pose, relative_to_robot=False):
...
+                the object to stop at - this is the distance between the origins so
+                e.g. from between Cozmo's 2 front wheels to the center of a cube
+                is ~40mm
+        Returns:
+            A :class:`cozmo.robot.GoToObject` action object which can be queried
+                to see when it is complete.
+        '''
+        if not isinstance(object, objects.ObservableObject):
+            raise TypeError("Invalid charger supplied")
```
Comment (gwatts) [inline src/cozmo/robot.py]: Presumably don't mean charger here, since it can be any  observable object..

Perhaps "Target must be an observable object" instead?

File: src/cozmo/world.py (hunk 6 above)
Comment (MarkWesleyAtAnki) [inline src/cozmo/world.py]: I've always used TODO in the past, but seems that XXX is commonly used to point out odd but not necessarily broken (or to-be-done) things.

File: examples/drive_to_charger_test.py (hunk 8 above)
Comment (MarkWesleyAtAnki) [inline examples/drive_to_charger_test.py]: I will simplify, unfortunately at the cost of it working less reliably.

File: examples/drive_to_charger_test.py (hunk 1 above)
Comment (MarkWesleyAtAnki) [inline examples/drive_to_charger_test.py]: fixed

File: src/cozmo/objects.py (hunk 2 above)
Comment (MarkWesleyAtAnki) [inline src/cozmo/objects.py]: fixed

File: src/cozmo/objects.py (hunk 3 above)
Comment (MarkWesleyAtAnki) [inline src/cozmo/objects.py]: Was following the pattern from the other events, but yeah i guess they should be consistent. Fixed.

File: src/cozmo/objects.py (hunk 9 above)
Comment (MarkWesleyAtAnki) [inline src/cozmo/objects.py]: Fixed

File: src/cozmo/objects.py (hunk 10 above)
Comment (MarkWesleyAtAnki) [inline src/cozmo/objects.py]: Fixed

File: src/cozmo/robot.py (hunk 11 above)
Comment (MarkWesleyAtAnki) [inline src/cozmo/robot.py]: Fixed

File: src/cozmo/world.py (hunk 6 above)
Comment (gwatts) [inline src/cozmo/world.py]: Yeah `TODO` implies that there's actionable work to be done IMO; if you're bored and grep the source tree for it, you should find things  you can work on.

`XXX` means warning, might change, heads up, etc

File: src/cozmo/world.py [hunk 12]
```diff
@@ -102,6 +116,10 @@ def __init__(self, conn, robot, **kw):
 
         self.light_cubes = {}
 
+        #: :class:`cozmo.objects.Charger`: Cozmo's charger.
```
Comment (gwatts) [inline src/cozmo/world.py]: there is no `cozmo.objects.Charger` class... presumably mean `ObservableObject `

Comment (MarkWesleyAtAnki) [inline src/cozmo/world.py]: There is - I added cozmo.objects.Charger as part of this PR
